lib/http1.c

```diff
@@ -92,8 +92,17 @@ struct st_h2o_http1_chunked_entity_reader {
     struct phr_chunked_decoder decoder;
 };
 
+enum tunnel_type {
+    TUNNEL_TCP,
+    TUNNEL_UDP,
+};
+
 struct st_h2o_http1_tunnel_t {
-    h2o_httpclient_tunnel_t *server;
+    enum tunnel_type tunnel_type;
+    union {
+        h2o_httpclient_tunnel_t *tcp;
+        h2o_httpclient_udp_tunnel_t *udp;
+    } server;
```

Comment:
These changes makes me wonder if it is a good idea to introduce `h2o_httpclient_udp_tunnel_t` as something separate from `h2o_httpclient_tunnel_t`.

From the perspectives of HTTP/1 and HTTP/2 implementation, only "streams" can carry UDP datagrams. Even for HTTP/3, use of stream is the default; use of QUIC datagram frames is an "option." Regardless of the HTTP version being used, encapsulation format of UDP datagrams sent over the HTTP streams of respective versions is going to be the same.

Considering these aspects, would it make sense to merge the definion of `h2o_httpclient_udp_tunnel_t` into `h2o_httpclient_tunnel_t`, and letting the tunnel handle encapsulation?

I mean, h2o_httpclient_tunnel_t can have an optional `on_udp_read` callback. When it is set by the H3 server-side implementation, the tunnel would provide the UDP datagrams as they are through the callback. OTOH, when the callback is set to NULL, h2o_httpclient_tunnel_t would provide the encapsulated byte stream through the `on_read` callback.

For the write side, h2o_httpclient_tunnel_t can always provide both interfaces (i.e. `write_` and `udp_write`) assuming that a UDP tunnel is being opened. It would be the responsibility of the tunnel to unpack the encapsulated payload that it receives through the `write_` callback.

Reply:
that makes sense to me, thank you. What do you think of moving all the code (tcp or udp) to lib/common/tunnel.c?

Reply:
Mmm, i see now that it would make sense to keep both tunnels in their current files, there might be a handful of shared functions, but it'll be easier to split the implementations in different modules.

Reply:
I do not have a concrete opinion, but I am not sure how much code can be shared, based on my assumption that the buffering strategy would be different. In case of TCP, the tunnel cannot drop anything, while in case of UDP, the tunnel is expected to have a fixed size maximum.

I would be fine with splitting the two, especially when the amount of logic that can be shared would be less than the amount of the logic that could be shared on the user-side; our H1 and H3 server-side implementations both use the tunnel API, they do have some code that look similar, but I tend to think that not trying to share stuff is giving us more benefit than drawbacks.